0003-erl_child_setup_thread.patch

```diff
@@ -0,0 +1,1219 @@
++                }
++            }
++        }
++    } else {
++        warning("unable to execute command: \"%s\"\n", cmd);
++    }
++
++    if (!launched) {
++        close(pipes[0]);
++        close(pipes[1]);
++        close(pipes[2]);
++        free(pipes);
++    }
++    pthread_exit(0);
++  child_error:
++    fprintf(stderr, "erl_child_setup: failed with error %d on line %d\r\n",
++            errno, errln);
++    pthread_exit((void *)1);
++}
++
++void *erl_child_setup_thread(void *arg)
++{
++    int uds_fd = (long)arg;
++    DEBUG_PRINT("in %s, uds_fd %d", __func__, uds_fd);
++
++    while (1) {
++        /* shouldn't really need select, but there might be something else to deal with... */
++        fd_set read_fds;
++        int res;
++        FD_ZERO(&read_fds);
++        FD_SET(uds_fd, &read_fds);
++        DEBUG_PRINT("child_setup selecting on %d", uds_fd);
++        res = select(uds_fd+1, &read_fds, NULL, NULL, NULL);
++
++        if (res < 0) {
++            if (errno == EINTR) continue;
++            ABORT("Select failed: %d (%d)",res, errno);
++        }
++
++        if (FD_ISSET(uds_fd, &read_fds)) {
++            int res;
++            int *pipes;
++            ErtsSysForkerProto proto;
++            pthread_t new_child;
++
++            errno = 0;
++            pipes = malloc(sizeof(int) * 3);
++            if ((res = sys_uds_read(uds_fd, (char*)&proto, sizeof(proto),
++                                    pipes, 3, MSG_DONTWAIT)) < 0) {
++                if (errno == EINTR)
++                    continue;
++                DEBUG_PRINT("erl_child_setup failed to read from uds: %d, %d", res, errno);
++                goto fail;
++            }
++
++            /* This will only work in Nanos and is likely to break if run
++               under Linux. Nanos currently does not have support for passing
++               file descriptors via SCM_RIGHTS, let alone even supporting
++               ancillary data over a unix domain socket. We left the sys_uds_*
++               interface alone, and still pass pipes to the read above, but
++               all it will do in Nanos is fill it with zeros. Thankfully, the
++               original file descriptors are stashed in proto, so just copy
++               them back out here. */
++            memcpy(pipes, proto.u.start.fds, sizeof(int) * 3);
++
++            if (res == 0) {
++                DEBUG_PRINT("uds was closed!");
++                goto fail;
++            }
++            DEBUG_PRINT("### proto action %d, pipes[0] %d, pipes[1] %d",
++                        proto.action, pipes[0], pipes[1]);
++            /* Since we use unix domain sockets and send the entire data in
++               one go we *should* get the entire payload at once. */
++            ASSERT(res == sizeof(proto));
++            ASSERT(proto.action == ErtsSysForkerProtoAction_Start);
++
++            if (pthread_create(&new_child, 0, start_new_child, pipes) < 0) {
```

Comment:
Since this thread is not joined, it should be detached (by calling pthread_detach()), otherwise when it terminates its resources are not released and there will be a memory leak.

Reply:
Yes, thanks. pthread_detach() calls have been added for all created threads.